### PR TITLE
feat(upgrade): support engine suspension and target switchover

### PR DESCRIPTION
**Pull request by derekbit**
#### Which issue(s) this PR fixes:

Issue Longhorn/longhorn#6001

### CODE REVIEW

**Review by DamiaSan — COMMENTED** [review]
I see you have implemented the flush command over nvme, but actually lvol bdev doesn't support that. I had implemented it, but this development caused a memory leak in the set external parent functionality, which we use in online rebuilding. I understood that we didn't need the flush over nvme and that we only used it at device mapper layer, so I discarded that commit in the new SPDK branch `longhorn-v24.05`. I thought to restart working over it, to remove the leak, after 1.7.0.

**Comment by derekbit** [comment]
> in online rebuilding. I understood that we didn't need the flush over nvme and that we only used it at device mapper layer, so I discarded that commit in the new SPDK branch `longhorn-v24.05`. I thought to restart working over it, to remove the leak, after 1.7.0.

Got it. I think the flush subcommand is fine because I didn't use in the suspension and switchover functionalities. The subcommand is just for making the nvme command more complete. WDYT?

**Comment by DamiaSan** [comment]
> > in online rebuilding. I understood that we didn't need the flush over nvme and that we only used it at device mapper layer, so I discarded that commit in the new SPDK branch `longhorn-v24.05`. I thought to restart working over it, to remove the leak, after 1.7.0.
> 
> Got it. I think the flush subcommand is fine because I didn't use in the suspension and switchover functionalities. The subcommand is just for making the nvme command more complete. WDYT?

Ok, this is fine for now. In the future I will work again on the flush implementation so we will have it again.

**Review by DamiaSan — APPROVED** [review]
Just a couple of questions, but globally it LGTM

File: pkg/nvme/initiator.go [hunk 1]
```diff
@@ -161,42 +285,48 @@ func (i *Initiator) Start(transportAddress, transportServiceID string, needDmDev
...
 	if i.ControllerName == "" {
-		return dmDeviceBusy, fmt.Errorf("failed to start NVMe initiator %s within %d * %v sec retries", i.Name, RetryCounts, RetryInterval.Seconds())
+		return dmDeviceBusy, fmt.Errorf("failed to start NVMe initiator %s within %d * %v sec retries", i.Name, maxNumRetries, retryInterval.Seconds())
 	}
 
-	for t := 0; t < int(waitDeviceTimeout.Seconds()); t++ {
-		if err = i.loadNVMeDeviceInfoWithoutLock(); err == nil {
+	for r := 0; r < maxNumWaitDeviceRetries; r++ {
+		err = i.loadNVMeDeviceInfoWithoutLock()
+		if err == nil {
```
[the remark this answers is not in the document]
Comment (derekbit) [reply pkg/nvme/initiator.go:315]: It is just my personal convention.

Yes, we can write:
```
if err = i.loadNVMeDeviceInfoWithoutLock(); err == nil {
	return nil
}
```
but err is a returned value here. I'm not comfortable with this method.

I prefer to put them on the same line when the scope is only within the check logic:
```
if err := i.loadNVMeDeviceInfoWithoutLock(); err == nil {
	return nil
}
```

File: pkg/nvme/initiator.go [hunk 2]
```diff
@@ -88,16 +171,53 @@ func (i *Initiator) Suspend(noflush, nolockfs bool) error {
...
+func (i *Initiator) Resume() error {
+	if i.hostProc != "" {
+		lock := nsfilelock.NewLockWithTimeout(util.GetHostNamespacePath(i.hostProc), LockFile, LockTimeout)
+		if err := lock.Lock(); err != nil {
+			return errors.Wrapf(err, "failed to get file lock for initiator %s", i.Name)
+		}
+		defer lock.Unlock()
+	}
+
+	if err := i.resumeLinearDmDevice(); err != nil {
```
[the remark this answers is not in the document]
Comment (derekbit) [reply pkg/nvme/initiator.go:198]: Yes, we can check it, but resuming an active won't raise an error. Should we check it?

[the remark this answers is not in the document]
Comment (DamiaSan) [reply pkg/nvme/initiator.go:198]: Ok then

File: pkg/nvme/initiator.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (DamiaSan) [reply pkg/nvme/initiator.go:315]: Ok thanks

**Comment by derekbit** [comment]
@DamiaSan I appended an extra commit for resolving build environment error. Can you review it again? Thank you.

**Review by DamiaSan — DISMISSED** [review]

**Comment by derekbit** [comment]
@DamiaSan Can you help review the latest commit? Thank you.

**Review by DamiaSan — APPROVED** [review]

**Pull request merged**